src/strkey.js

```diff
@@ -136,31 +146,117 @@ export class StrKey {
+   * @param   {string} address  address to decode
    * @returns {Buffer}
    */
-  static decodeSha256Hash(data) {
-    return decodeCheck('sha256Hash', data);
+  static decodeSignedPayload(address) {
+    return decodeCheck('signedPayload', address);
+  }
+
+  /**
+   * Checks validity of alleged signed payload (P...) strkey address.
+   * @param   {string} address  signer key to check
+   * @returns {boolean}
+   */
+  static isValidSignedPayload(address) {
+    return isValid('signedPayload', address);
+  }
+
+  static getVersionByteForPrefix(address) {
+    return strkeyTypes[address[0]];
   }
 }
 
-// Warning: This isn't a *definitive* check of validity, but rather just a
-// basic-effort check.
+/**
+ * Sanity-checks whether or not a strkey *appears* valid.
+ *
+ * @param  {string}  versionByteName the type of strkey to expect in `encoded`
+ * @param  {string}  encoded         the strkey to validate
+ *
+ * @return {Boolean} whether or not the `encoded` strkey appears valid for the
+ *     `versionByteName` strkey type (see `versionBytes`, above).
+ *
+ * @note This isn't a *definitive* check of validity, but rather a best-effort
+ *     check based on (a) input length, (b) whether or not it can be decoded,
+ *     and (c) output length.
+ */
 function isValid(versionByteName, encoded) {
-  // it's either non-muxed && len=56, or muxed && len=69
-  if (encoded && encoded.length !== 56 && encoded.length !== 69) {
+  if (!isString(encoded)) {
     return false;
   }
 
-  try {
-    const decoded = decodeCheck(versionByteName, encoded);
-    if (decoded.length !== 32 && decoded.length !== 40) {
+  // basic length checks on the strkey lengths
+  switch (versionByteName) {
+    case 'ed25519PublicKey': // falls through
+    case 'ed25519SecretSeed': // falls through
+    case 'preAuthTx': // falls through
+    case 'sha256Hash':
+      if (encoded.length !== 56) {
+        return false;
+      }
+      break;
+
+    case 'med25519PublicKey':
+      if (encoded.length !== 69) {
+        return false;
+      }
+      break;
+
+    case 'signedPayload':
+      if (encoded.length < 56 || encoded.length > 165) {
+        return false;
+      }
+      break;
+
+    default:
       return false;
-    }
+  }
+
+  let decoded = '';
+  try {
+    decoded = decodeCheck(versionByteName, encoded);
   } catch (err) {
     return false;
   }
-  return true;
+
+  // basic length checks on the resulting buffer sizes
+  switch (versionByteName) {
+    case 'ed25519PublicKey': // falls through
+    case 'ed25519SecretSeed': // falls through
+    case 'preAuthTx': // falls through
+    case 'sha256Hash':
+      return decoded.length === 32;
+
+    case 'med25519PublicKey':
+      return decoded.length === 40; // +8 bytes for the ID
+
+    case 'signedPayload':
+      return (
+        // 32 for the signer, +4 for the payload size, +64 for the max payload
+        decoded.length >= 32 + 4 && decoded.length <= 32 + 4 + 64
```

Comment:
According to the description in the latest CAP-40, the length of the payload is at least 1, maybe this should be modified to `decoded.length >= 32 + 4 + 4`?

See [Transaction Validity](https://github.com/stellar/stellar-protocol/blob/master/core/cap-0040.md#transaction-validity)

Reply:
This condition was only just introduced. Thanks for the catch!